code/modules/atmospherics/FEA_system.dm

```diff
@@ -1,322 +1,280 @@
-							LAZYLISTINIT(possible_space_borders)
-							possible_space_borders |= test
-							test.length_space_border++
-
-				if(test.length_space_border > 0)
-					possible_space_length += test.length_space_border
-				possible_members -= test
-
-		if(members.len > 1)
-			var/datum/air_group/group = new
-			if(possible_borders && (possible_borders.len > 0))
-				group.borders = possible_borders
-			if(possible_space_borders && (possible_space_borders.len > 0))
-				group.space_borders = possible_space_borders
-				group.length_space_border = possible_space_length
-
-			for(var/turf/simulated/test as anything in members)
-				test.parent = group
-				test.processing = 0
-				active_singletons -= test
-
-				test.dist_to_space = null
-				var/dist
-				for(var/P in possible_space_borders)
-					var/turf/simulated/b = P
-					if (b == test)
-						test.dist_to_space = 1
-						break
-					dist = GET_DIST(b, test)
-					if ((test.dist_to_space == null) || (dist < test.dist_to_space))
-						test.dist_to_space = dist
-
-			// Allow groups to determine if group processing is applicable after FEA setup
-			if(current_cycle) group.group_processing = FALSE
-			group.members = members
-			air_groups += group
-
-			group.update_group_from_tiles() //Initialize air group variables
-			return group
-		else
-			base.processing = 0 //singletons at startup are technically unconnected anyway
-			base.parent = null
-
-			if(base.air && base.air.check_tile_graphic())
-				base.update_visuals(base.air)
-
-		return null
-
-	process()
-		current_cycle++
-
-		process_tiles_to_space()
-		is_busy = TRUE
-
-		if(!explosions.exploding)
-			if(groups_to_rebuild.len > 0)
-				process_rebuild_select_groups()
-			LAGCHECK(LAG_REALTIME)
+	var/turf/space/space_sample //instead of repeatedly using locate() to find space, we should just cache a space tile ok
+
+/// Updates cached space sample if need be.
+// Returns: New space sample.
+/datum/controller/air_system/proc/update_space_sample()
+	if (!space_sample || !(space_sample.turf_flags & CAN_BE_SPACE_SAMPLE))
+		space_sample = locate(/turf/space)
+	return space_sample
+
+/datum/controller/air_system/proc/setup(datum/controller/process/air_system/controller)
+	parent_controller = controller
+
+	#if SKIP_FEA_SETUP == 1
+	return
+	#else
+
+	boutput(world, "<span class='alert'>Processing Geometry...</span>")
+
+	var/start_time = world.timeofday
+
+	for(var/turf/simulated/S in world)
+		if(!S.gas_impermeable && !S.parent)
+			assemble_group_turf(S)
+		S.update_air_properties()
+
+	boutput(world, "<span class='alert'>Geometry processed in [(world.timeofday-start_time)/10] seconds!</span>")
+	#endif
+
+/// Collects turfs into groups.
+/datum/controller/air_system/proc/assemble_group_turf(turf/simulated/base)
+	set waitfor = 0
+	var/list/turf/simulated/members = list(base) // Confirmed group members
+	var/list/turf/simulated/possible_members = list(base) // Possible places for group expansion
```

Comment:
maybe make these comments doc comments

Reply:
no can do, these are temp vars since they in a proc.

Reply:
nevermind then